### DISCUSSION TITLE
Can't get two I2C devices (each on their own bus) to work simultaneously (ESP32)

### DISCUSSION THREAD

**Discussion by Sum-Guy** [arendst/Tasmota#14580] [Q&A] [answered]
I have a BMP280 and AHT10.  Both are wired as I2C.  The AHT10 is connected to I/O 4 and 5 (SDA/SCL).  The BMP280 is connected to I/O 18 and 23 (SDA/SCL).  The AHT10 can't share the same I2C bus with other devices.  This is on a Wemos LoLin32 Lite Esp32 Development Board. 

I used the Gitpod.io Tasmo compiler to compile tasmota 10.1.0 for ESP32, i selected a few options (Temp/Hum sensor) and I also added the custom parameter lines for the AHT1x.

#ifndef USE_AHT1x
#define USE_AHT1x       // [I2cDriver43] Enable AHT10/15 humidity and temperature sensor (I2C address 0x38) (+0k8 code)
#endif

From the boot log:

00:00:00.002 HDW: ESP32-D0WDQ6
00:00:00.167 Project tasmota - Tasmota Version 10.1.0(TasmoCompiler-esp32generic)-2_0_1_1(2022-01-22T06:06:45)

When configuring the ESP32 using the web interface I can set the relevant I/O lines as SDA and SCL, and I notice beside each one is a small drop-down selector where I can select either "1" or "2".  I can select 1 or 2 (and nothing else) when only one or both sensors are connected.  Some examples I see of configuring these sensors shows (5) and (6) which I assume are the numbers I see in the drop-down menu but I never see "5" or "6" - only 1 or 2.

In any case, when I alternate setting each sensor to either 1 or 2 and go back to the main Tasmota page, I can see the sensor data properly for only 1 of the devices, not both.  I can sometimes see the BMP or the AHT but not both at the same time. 

In one case, the log shows this:

00:00:00.151 SHT: Sensor did not ACK command
00:00:00.212 I2C: AHT1X found at 0x38

In the other case, the log shows this:

00:00:00.179 SHT: Sensor did not ACK command
00:00:00.201 I2C: BMP280 found at 0x76

Can anyone explain what's going on here? 

Bonus question:  How many physical I2C interfaces can the ESP32 have?  Some people say only 2, that's why I see only (1) or (2) in the GUI module configuration I2C SDA/SCL drop-down box.  But everything I read about the ESP32 says that every I/O pin can be configured as I2C, so I'm confused about that.

**Comment by sfromis** [comment, marked as answer]
The Tasmota support for two I2C busses on ESP32 is not general purpose, but made for a specific setup. You cannot just put devices on both busses, and expect it to work, the infrastructure of Tasmota sensor handling is not built for that.

And AHT10 is a *pig* on the I2C bus; known to break bus communication if not alone on the bus. Docs from them even mentions that it should be alone on a bus. Bad sensor, best avoided (unless you want nothing else on the bus, of course).

ESP32 has support for two hardware I2C busses, unless it is ESP32-C3 having only one. Due to a multiplexer those busses are not tied to specific pins, but can move around to configured pins, within supported limits on pin usage. For general purpose, Tasmota only supports one.

**Comment by Sum-Guy** [comment]
Ok, so the I2C buses can be assigned to many different I/O pins, but there can only be 2 defined I2C buses, and that is a hardware limitation inside the ESP32. 

Is this the same situation for the ESP8266?

Regarding my situation, I am not defining more than 2 buses.  I bought a few of the AHT10's because they were cheap. 

But still, shouldn't my configuration work?

**Reply by sfromis** [reply]
Not with two busses.
> You cannot just put devices on both busses, and expect it to work, the infrastructure of Tasmota sensor handling is not built for that.

And AHT10 needs to be alone on one bus, as it behaves like a pig.

On ESP8266, there is no hardware I2C, this is implemented in software instead. In principle, it would be possible with a variable number of software I2C, but I do not think that this is supported by Arduino Core, which Tasmota uses. In principle, it would also be possible with software I2C on ESP32, in addition to the hardware I2C support, but again, this would take development work to support, and the general assumption is that with the I2C bus architecture supporting several ("many") *well-behaved* devices on a single bus, work to create full Tasmota support for more than one is not very likely to be prioritized.

TLDR: With Tasmota you can expect to use just one I2C bus for "general purpose", both on ESP8266 and ESP32, irrespective of what would be possible with the hardware. If you use Berry on ESP32 to code your own I2C driver(s), I think that this could take advantage of the 2nd existing bus.

**Comment by Sum-Guy** [comment]
Wait a minute. I thought I saw examples where you can have several I2C devices sharing the same SDA pin, but using separate SCL pins (or maybe I'm mixing up the SDA/SCL) but you are creating "virtual" I2C that way?  And Tasmota can do that -and have more than 2 of those?  But still, I'm not doing that, I have 2 completely separate I2C buses, which is not violating any rule I think.  ?

**Reply by barbudor** [reply]
The problem is that most I2C drivers have been written prior to adding this capability for 2x I2C buses
This has been specifically introduced for the M5Stack Core2 which has an internal sensor connected to a private, on-board, I2C bus.
The driver for this sensor will work only wiht IC2 bus #2 
All other Tasmota sensors will work exclusively with I2C bus #1

It's not that it couldn't be done
Is that someone need to take the HUGE job of rewritting EVERY I2C drivers to support multiple buses.
No one has taken it yet.

**Reply by sfromis** [reply]
While it may be technically possible to play tricks with the I2C bus to *fake* multiple busses, including using bus multiplexers, this is just not supported by the Tasmota drivers coded on the assumption that there is just *one* bus, with all the available devices.

**Comment by Sum-Guy** [comment]
So operating an ESP32 (or even 8266) with 2 independent I2C buses is basically never done?  Because it is fundamentally unreliable as currently implimented in Tasmota?

**Reply by techman83** [reply]
Not unreliable. The [work](https://github.com/techman83/Tasmota/blob/development/CHANGELOG.md#9311-20210320) for 2 buses has been done, but most of the drivers were written before this work was done. So it would take refactoring all I2C drivers to harness this.

I had a look (I had a need for 2 buses as I'm abusing I2C a little), and I opted to maintain a branch that [patches](https://github.com/techman83/Tasmota/commit/8d3705cdc83dd0e5fb4e8251b1fa1398aee5e9e2) the specific driver I needed onto the other bus. Which works just fine.

**Comment by Sum-Guy** [comment]
So I'd be looking at the driver (for either the BMP280 or AHT10) and changing all occurrences of "wire" to wire1?  Or maybe wire2 if setting the device to use the second I2C bus?

**Reply by techman83** [reply]
Wire is bus 1, wire1 is bus 2. And if that's how the driver is written, it'll probably work. The [MPU driver](https://github.com/ardnest/Tasmota/blob/development/tasmota/xsns_85_mpu6886.ino#L73) looks to have bus detection as part of the initialisation of the driver.

My approach was definitely a hack, so YMMV.

**Reply by sfromis** [reply]
Still beware of AHT10 crapping on the bus, expect high risk of trouble if this sensor is now allowed to be alone on one I2C bus.
